File: ds/two_sum_hash.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

bool check(int *arr, int sum, int n);

int main()
{
    int n, sum = 0;
    cout << "Length of arr";
    cin >> n;
    int arr[n];
    cout << "\nEnter array elements";
    for (int i = 0; i < n; i++)
    {
        int x;
        cin >> x;
        arr[i] = x;
    }
    cout << "\nEnter sum to check for ";
    cin >> sum;

    bool present = check(arr, sum, n);
    if (present)
        cout << "true";
    else
        cout << "false";
}

bool check(int *arr, int sum, int n)
{
    unordered_map<int, int> m;
    for (int i = 0; i < n; i++)
    {
        int complement = sum - *(arr + i);
        if (m.find(complement) == m.end())
            m[*(arr + i)] = i;
        else
            return true;
    }
    return false;
}

File: ds/two_sum_hash.py
def check_for_sum(arr, sumc):
    map = {}
    for key, i in enumerate(arr):
        map[key] = 0
    for key in map.keys():
        complement = sumc - key
        if complement in map.keys():
            return True
    return False


def main():
    n = int(input("Enter size of list"))
    arr = list()
    for i in range(n):
        x = int(input("enter element "))
        arr.append(x)

    sumc = int(input("enter sum to check for "))
    exists = check_for_sum(arr, sumc)
    if exists:
        print("true")
    else:
        print("false")


if __name__ == "__main__":
    main()
